### PR TITLE
[SPARK-27448][SQL] File source V2 table provider should be compatible with V1 provider

### PR DESCRIPTION
**Pull request by gengliangwang**
## What changes were proposed in this pull request?

In the rule `PreprocessTableCreation`, if an existing table is appended with a different provider, the action will fail.
Currently, there are two implementations for file sources and creating a table with file source V2 will always fall back to V1 FileFormat. We should consider the following cases as valid:
1. Appending a table with file source V2 provider using the v1 file format
2. Appending a table with v1 file format provider using file source V2 format

## How was this patch tested?

Unit test

### CODE REVIEW

**Comment by gengliangwang** [comment]
The issue is found in `MetastoreDataSourcesSuite` while migrating Parquet to file data source V2. It blocks the migration. We should
1. allow such scenarios and consider the v1/v2 providers as compatible
 or
2.change the existing tests.

Option 1 seems more reasonable.

**Comment by gengliangwang** [comment]
@cloud-fan

**Comment by gengliangwang** [comment]
retest this please.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/rules.scala [hunk 1]
```diff
@@ -235,6 +236,24 @@ case class PreprocessTableCreation(sparkSession: SparkSession) extends Rule[Logi
       }
   }
 
+  private def compatibleProviders(a: Class[_], b: Class[_]): Boolean = {
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/rules.scala]: how about
```
def fallBackToV1(cls: Class[_]) = cls.newInstance match {
  case f: FileDataSourceV2 => f.fallbackFileFormat
  case _ => cls
}

...

fallBackToV1(existingProvider) == fallBackToV1(specifiedProvider)
```

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Pull request closed without merging**